File: Domain-Module/User-Module/User-Infrastructure/src/main/kotlin/com/pawith/userinfrastructure/repository/UserRepository.java
package com.pawith.userinfrastructure.repository;

import com.pawith.userinfrastructure.entity.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.List;

public interface UserRepository extends JpaRepository<User, Long> {
    User findByEmail(String email);
    boolean existsByEmail(String email);

    @Query("select u from User u where u.id in :ids")
    List<User> findAllByIds(List<Long> ids);

    @Query("select u from User u where u.nickname like %:nickname% and u.id in :ids")
    List<User> findAllByNicknameAndIds(String nickname, List<Long> ids);

}

File: Domain-Module/src/main/kotlin/com/pawith/domain/user/repository/UserRepository.kt
package com.pawith.domain.user.repository

import com.pawith.domain.user.User
interface UserRepository {
    fun findByEmailOrNull(email: String): User?
    fun findByIdOrNull(id: Long): User?
    fun existsByEmail(email: String): Boolean

    fun findAllByIds(ids: List<Long>): List<User>

    fun findAllByNicknameAndIds(nickname: String, ids: List<Long>): List<User>

    fun save(user: User) : User

    fun update(user: User) : User

    fun deleteById(id: Long)
}